### PR TITLE
Sim2h use IM crate for state

### PR DESCRIPTION
**Pull request by neonphog**
Switch to using IM for sim2h state in preparation for cloning the state to hand off to parallel handle_message jobs

### CODE REVIEW

File: crates/sim2h/src/cache.rs [hunk 1]
```diff
@@ -25,6 +25,27 @@ pub struct Space {
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Space")
+            .field("agents", &self.agents)
+            .field("all_aspect_hashes", &self.all_aspects_hashes)
+            .field("missing_aspects", &self.missing_aspects)
+            .finish()
+    }
+}
+
+impl Clone for Space {
```
Comment (struktured) [inline crates/sim2h/src/cache.rs]: What is the difference between this and deriving `Clone`?

File: crates/sim2h/src/lib.rs [hunk 2]
```diff
@@ -349,11 +356,11 @@ impl Sim2h {
...
 
         if let Some(ConnectionState::Joined(space_address, agent_id)) =
-            self.connection_states.write().remove(uri)
+            self.state.connections.remove(uri)
         {
-            if let Some(space_lock) = self.spaces.get(&space_address) {
-                if space_lock.write().remove_agent(&agent_id) == 0 {
-                    self.spaces.remove(&space_address);
+            if let Some(space) = self.state.spaces.get_mut(&space_address) {
+                if space.remove_agent(&agent_id) == 0 {
```
Comment (struktured) [inline crates/sim2h/src/lib.rs]: Is this a modification of the original or a cloned copy?

Comment (struktured) [reply crates/sim2h/src/lib.rs]: Nevermind - I see that there's a clone here at all.

File: crates/sim2h/src/cache.rs (hunk 1 above)
Comment (neonphog) [reply crates/sim2h/src/cache.rs]: the derive generates code very similar to this... doing it manually lets me do the `box_clone` on the crypto system

File: crates/sim2h/src/lib.rs (hunk 2 above)
Comment (neonphog) [reply crates/sim2h/src/lib.rs]: I haven't put any code in to clone yet, so all the mut calls modify the original

File: crates/sim2h/src/cache.rs (hunk 1 above)
Comment (AshantiMutinta) [reply crates/sim2h/src/cache.rs]: There is a rust crate  if we ever wanted to go that route :

https://github.com/dtolnay/dyn-clone

**Comment by neonphog** [comment]
closing - these changes will be included in a later PR

**Pull request closed without merging**